### PR TITLE
Allow pods to define multiple environment variables from a whole ConfigMap

### LINKED ISSUE #26299: Allow auto-population of environment variables from ConfigMap
The current approach for populating environment variables from ConfigMaps ends up pretty verbose. With the following ConfigMap:

```
apiVersion: v1
data:
  space-ships: 1
  ship-type: battle-cruiser
  weapon: laser-cannon
kind: ConfigMap
metadata:
  name: space-config
```

I have to manually expose each variable separately:

```
apiVersion: v1
kind: Pod
metadata:
  name: space-simulator
spec:
  containers:
  - name: space-simulator
    image: foo/space-simulator
    env:
    - name: SPACE_SHIPS
      valueFrom:
        configMapKeyRef:
          name: space-config
          key: space-ships
    - name: SHIP_TYPE
      valueFrom:
        configMapKeyRef:
          name: space-config
          key: ship-type
    - name: WEAPON
      valueFrom:
        configMapKeyRef:
          name: space-configs
          key: weapon
```

This seems unnecessary and error-prone if all I want to express is that the pod should pull all the data from the ConfigMap and populate environment variables from them.

It would be nice if this could be done in more concise manner, perhaps by introducing an `envFrom` field. For example:

```
apiVersion: v1
kind: Pod
metadata:
  name: space-simulator
spec:
  containers:
  - name: space-simulator
    image: foo/space-simulator
  - envFrom:
      configMap: space-config
```

The idea being that `envFrom` could support other ways of auto-population, e.g. secrets.

### PR DESCRIPTION
**Pull request by fraenkel**
Allow environment variables to be populated from ConfigMaps

- ConfigMaps represent an entire set of EnvVars
- EnvVars can override ConfigMaps

fixes #26299

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36245)

### CODE REVIEW

**Comment by dims** [comment]
@k8s-bot ok to test

**Comment by fraenkel** [comment]
The failure is not related to this change.
`
W1110 03:33:33.283] Error from server (NotFound): replicasets "nginx-618515232" not found
`

If I dump all replicasets, I see
```
NAMESPACE     NAME               DESIRED   CURRENT   READY     AGE
default       nginx-3347634067   2         2         0         4s
default       nginx-4160935279   2         2         0         3s
```

File: pkg/api/v1/types.go [hunk 1]
```diff
@@ -1478,6 +1484,10 @@ type Container struct {
 	// Default is false.
 	// +optional
 	TTY bool `json:"tty,omitempty" protobuf:"varint,18,opt,name=tty"`
+	// List of environment variables to set in the container.
```
Comment (thockin) [inline pkg/api/v1/types.go]: This needs better explanation about priority wrt named `env` values (I assume they override this) and about ordering (I assume later entries override earlier entries), and about how configmap keys get transformed into env var names.  Keys in a configmap can hold values that env vars can not - what happens?

Comment (fraenkel) [inline pkg/api/v1/types.go]: 1. Add validation of keys.
2. Add the precedence ordering which should be order of precedence Env, EnvFrom, Services.
3. Add support for expansion of Envs with a runtimeVal.  I noticed there weren't (m)any tests for it.

**Comment by thockin** [comment]
I'm stealing this PR since Daniel is out and I have been looking at alternative ideas (and because I want something like this for my own use :).

@dhoer does this satisfy?

Please ping this thread when you have updated in response to comments.

**Comment by dhoer** [comment]
@thockin Yes, but see https://github.com/kubernetes/kubernetes/issues/8800#issuecomment-261308901

**Comment by fraenkel** [comment]
@thockin Updated the code. PTAL.

**Comment by bgrant0607** [comment]
cc @pmorie

**Comment by bgrant0607** [comment]
cc @pwittrock

File: pkg/api/v1/types.go [hunk 2]
```diff
@@ -1478,6 +1484,14 @@ type Container struct {
 	// Default is false.
 	// +optional
 	TTY bool `json:"tty,omitempty" protobuf:"varint,18,opt,name=tty"`
+	// List of sources to populate environment variables in the container.
```
Comment (bgrant0607) [inline pkg/api/v1/types.go]: Please put this just below Env.

File: pkg/api/v1/types.go [hunk 3]
```diff
@@ -1233,6 +1233,12 @@ type SecretKeySelector struct {
 	Key string `json:"key" protobuf:"bytes,2,opt,name=key"`
 }
 
+// EnvFromSource represents the source of a set of EnvVars
+type EnvFromSource struct {
+	// The ConfigMap to select from
+	ConfigMap *LocalObjectReference `json:"configMap,omitempty" protobuf:"bytes,1,opt,name=configMap"`
```
Comment (bgrant0607) [inline pkg/api/v1/types.go]: The convention we normally use is fooRef. configMapRef.

File: pkg/api/v1/types.go [hunk 4]
```diff
@@ -1478,6 +1484,14 @@ type Container struct {
 	// Default is false.
 	// +optional
 	TTY bool `json:"tty,omitempty" protobuf:"varint,18,opt,name=tty"`
+	// List of sources to populate environment variables in the container.
+	// The keys defined within a source must be a C_IDENTIFIER.
+	// When a key exists in multiple sources, the value associated with the last
+	// source will take precedence.
+	// All EnvVars will take precedence over any listed source.
```
Comment (bgrant0607) [inline pkg/api/v1/types.go]: We generate user-facing documentation from this file, so please use json field names and plain English rather than Go type and field names.

**Review by bgrant0607 — DISMISSED** [review]

**Comment by bgrant0607** [comment]
@fraenkel This needs a better release note. However, I see that you deleted the PR template, so you'll need to copy the release-note text back into the PR description. Please describe the feature in more detail (e.g., the json field name).

Also, this will need user-facing documentation in kubernetes.github.io before it can go into a release. Please ask kubernetes-sig-docs at googlegroups.com where the best place to put the documentation would be.

**Comment by pmorie** [comment]
@fraenkel 

I don't want to throw a blanket on this at all, but this is the kind of API change I would strongly prefer to see a proposal for *before* we debate merits of PR that changes code.  I am not asking you to change this PR at all, necessarily, but I would really like to see a clear, proposal-like description of what the API change is so that we can discuss the specifics and possible corner cases before going much further.

Also, we could really use the same functionality for secrets, so I would like to handle both of these resources in the same release (though probably not the same PR).

I have not actually read the API changes here yet, so no comment on the content of this PR.

**Comment by pmorie** [comment]
@fraenkel In case it's not clear from other issues I have commented on, let me go on record as saying I am quite happy that you picked this up, it has needed doing for a while.  I just want a doc change I can process before we start reviewing a 30 file PR.

**Comment by fraenkel** [comment]
@pmorie If only API changes were small, 20 generated files adds up.

A container can now specify a list of ConfigMaps that will be included as part of the environment. The order of precedence for environment variables are env (env vars), envFrom (config maps), and then services.
Variable references defined by env.Value are resolved via variables defined by env, envFrom and services.
Validation of ConfigMap keys is done during environment variable processing. An error is returned if any key does not conform to a "C" identifier.

The describe command will display the configmap name that have been defined as part of the environment variable section.

**Comment by fraenkel** [comment]
@bgrant0607 Changes have been made. Tried my best to tweak the docs.

File: pkg/api/v1/types.go [hunk 5]
```diff
@@ -1413,6 +1419,14 @@ type Container struct {
...
 	// +optional
 	Env []EnvVar `json:"env,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,7,rep,name=env"`
+	// List of sources to populate environment variables in the container.
+	// The keys defined within a source must be a C_IDENTIFIER.
+	// When a key exists in multiple sources, the value associated with the last
+	// source will take precedence.
+	// All env values will take precedence over any listed source.
+	// Cannot be updated.
+	// +optional
+	EnvFrom []EnvFromSource `json:"envFrom,omitempty" protobuf:"bytes,19,rep,name=envFrom"`
```
Comment (bgrant0607) [inline pkg/api/v1/types.go]: Hmm. We'd need to extend strategic merge patch in order to support this properly. We definitely want automated processes to be able to add additional configmaps (#17097).

One alternative would be to not use the OneOf pattern in EnvFromSource and instead include an object reference that would specify the resource type (kind and apiVersion). Another alternative would be to use separate arrays for configmaps, secrets, services, and any types we add in the future that would define multiple environment variables. The former sounds less objectionable.

Any other ideas @pwittrock?

**Comment by fraenkel** [comment]
Just to keep everyone on the same page. I plan to create a proposal to further discuss the direction of this. The goal will be to decide on the behavior of the configmap processing and how it should be defined within the ContainerSpec.

**Comment by pmorie** [comment]
@fraenkel 

Thanks for that -- my primary concerns that I would like to work through in such a proposal are:

1.  How do environment vars get manifested into a container from this API feature, given some example pods and config maps?
2.  How are collisions handled when multiple configmaps are referred to that have the same keys?
3.  How are collisions handled when a secret and a configmap with the same key are referred to?
4.  How can I customize (within reason) the manifestation of env vars into a pod spec that uses this feature?
5.  How can I tell from the pod's status which env vars I actually got?

Hopefully that provides some good food for thought.

**Comment by fraenkel** [comment]
Finally submitted the proposal under #37295 
One difference is the introduction of a prefix for all keys from a Source which is only a ConfigMap thus far.

**Comment by fraenkel** [comment]
@k8s-bot node e2e test this

**Comment by dims** [comment]
@k8s-bot node e2e test this

**Comment by thockin** [comment]
@k8s-bot <https://github.com/k8s-bot> node e2e test this

**Comment by fraenkel** [comment]
Depending on how we want to support optional ConfigMaps, the ConfigMapRef may need to be an actual struct.

File: pkg/kubelet/kubelet_pods.go [hunk 6]
```diff
@@ -414,6 +414,44 @@ func (kl *Kubelet) makeEnvironmentVariables(pod *v1.Pod, container *v1.Container
 		return result, err
 	}
 
+	var (
+		configMaps = make(map[string]*v1.ConfigMap)
+		tmpEnv     = make(map[string]string)
+	)
+	for _, envFrom := range container.EnvFrom {
```
Comment (pmorie) [inline pkg/kubelet/kubelet_pods.go]: Add a comment here explaining the order of evaluation of `EnvFrom` and `Env`

File: pkg/kubelet/kubelet_pods.go [hunk 7]
```diff
@@ -414,6 +414,44 @@ func (kl *Kubelet) makeEnvironmentVariables(pod *v1.Pod, container *v1.Container
...
+		if envFrom.ConfigMapRef != nil {
+			name := envFrom.ConfigMapRef.Name
+			configMap, ok := configMaps[name]
+			if !ok {
+				if kl.kubeClient == nil {
```
Comment (pmorie) [inline pkg/kubelet/kubelet_pods.go]: I have doubts about whether this nil check is needed.  I realize that you're following the current pattern in this method, but as the original author of the method, I have doubts.

File: pkg/kubelet/kubelet_pods.go [hunk 8]
```diff
@@ -414,6 +414,44 @@ func (kl *Kubelet) makeEnvironmentVariables(pod *v1.Pod, container *v1.Container
...
+				}
+				configMap, err = kl.kubeClient.Core().ConfigMaps(pod.Namespace).Get(name, metav1.GetOptions{})
+
+				if err != nil {
+					return result, err
+				}
+				configMaps[name] = configMap
+			}
+			for k, v := range configMap.Data {
+				if errMsgs := utilvalidation.IsCIdentifier(k); len(errMsgs) != 0 {
```
Comment (pmorie) [inline pkg/kubelet/kubelet_pods.go]: I don't think this is the right behavior.  We shouldn't error out here -- skip this key or transform it.

File: pkg/kubelet/kubelet_pods_test.go [hunk 9]
```diff
@@ -605,6 +607,127 @@ func TestMakeEnvironmentVariables(t *testing.T) {
 				},
 			},
 		},
+		{
+			name: "configmap",
+			ns:   "test1",
+			container: &v1.Container{
+				Env: []v1.EnvVar{
```
Comment (pmorie) [inline pkg/kubelet/kubelet_pods_test.go]: nit, since `EnvFrom` is processed before `Env`, it makes tests of this functionality much more readable to me if `EnvFrom` appears before `Env`.

File: pkg/api/types.go [hunk 10]
```diff
@@ -1133,6 +1133,25 @@ type SecretKeySelector struct {
 	Key string
 }
 
+type EnvFromSource struct {
```
Comment (pmorie) [inline pkg/api/types.go]: Godoc for this type

File: pkg/api/types.go [hunk 11]
```diff
@@ -1276,6 +1295,8 @@ type Container struct {
 	Ports []ContainerPort
 	// +optional
 	Env []EnvVar
+	// +optional
```
Comment (pmorie) [inline pkg/api/types.go]: Godoc

File: pkg/api/v1/types.go [hunk 12]
```diff
@@ -1413,6 +1433,14 @@ type Container struct {
 	// Cannot be updated.
 	// +optional
 	Env []EnvVar `json:"env,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,7,rep,name=env"`
```
Comment (pmorie) [inline pkg/api/v1/types.go]: A couple things:

1.  Put `EnvFrom` before `Env` in this type
2.  Godoc for `Env` should incorporate info about overriding things set from `EnvFrom`

File: pkg/kubectl/describe.go [hunk 13]
```diff
@@ -1012,6 +1012,15 @@ func describeContainerEnvVars(container api.Container, resolverFn EnvVarResolver
 		none = "\t<none>"
 	}
 	w.Write(LEVEL_2, "Environment Variables:%s\n", none)
+
+	for _, e := range container.EnvFrom {
+		if len(e.Prefix) == 0 {
+			w.Write(LEVEL_3, "%s\tConfigMap\n", e.ConfigMapRef.Name)
```
Comment (pmorie) [inline pkg/kubectl/describe.go]: I think these need to be further qualified or broken out so that it's clear that all keys in the configmap are going to get injected.

File: test/e2e/common/configmap.go [hunk 14]
```diff
@@ -155,7 +155,7 @@ var _ = framework.KubeDescribe("ConfigMap", func() {
 
 	It("should be consumable via environment variable [Conformance]", func() {
```
Comment (pmorie) [inline test/e2e/common/configmap.go]: This should be a new test.  Keep the old test around and add a new one for `EnvFrom`

**Review by pmorie — COMMENTED** [review]
Generally looking good, noted a few nits.

File: pkg/kubelet/kubelet_pods.go (hunk 7 above)
Comment (fraenkel) [inline pkg/kubelet/kubelet_pods.go]: This type of check happens everywhere because there is no guarantee at the time of creation. It seems to allow a nil kubeClient.

File: pkg/kubelet/kubelet_pods.go (hunk 8 above)
Comment (fraenkel) [inline pkg/kubelet/kubelet_pods.go]: This is the behavior you get if you define invalid env vars as well. The problem here is that we need to do it at run time instead. But this is how the proposal was written.

We can't transform it since we don't know what was invalid. Skipping is even worse since you have no idea why an environment variable went missing.
Currently you will see your container failed to start and a message in the log.

File: pkg/api/types.go (hunk 11 above)
Comment (fraenkel) [inline pkg/api/types.go]: Everything is in v1/types. I was copying the style see EnvVar.

File: pkg/kubectl/describe.go (hunk 13 above)
Comment (fraenkel) [inline pkg/kubectl/describe.go]: Not sure how other than dumping the ConfigMap which may not exist.

File: pkg/api/v1/types.go (hunk 12 above)
Comment (fraenkel) [inline pkg/api/v1/types.go]: Seems redundant. EnvFrom already says Env will take precedence.

File: pkg/api/defaulting_test.go [hunk 15]
```diff
@@ -149,7 +149,7 @@ func TestDefaulting(t *testing.T) {
 				if !expectedChanged || changedOnce {
 					break
 				}
-				if iter > 200 {
+				if iter > 300 {
```
Comment (pmorie) [inline pkg/api/defaulting_test.go]: This should be a separate PR.

Comment (fraenkel) [inline pkg/api/defaulting_test.go]: I can but then it needs to go first or else these tests don't pass.

**Comment by pmorie** [comment]
@fraenkel the formula I recommend for PRs like this is three commits:

1.  A commit w/ API changes
2.  A commit with generated code
3.  A commit with other changes (ie, kubelet, e2es, etc)

Please squash out the 'feedback' commits, those don't have utility in master

File: pkg/kubelet/kubelet_pods.go (hunk 8 above)
Comment (thockin) [inline pkg/kubelet/kubelet_pods.go]: I agree - mutating it is wrong, ignoring it is wrong.

**Comment by pmorie** [comment]
I don't agree with one of the finer points of the behavior here, but it is as it was stated in the proposal, so we'll debate in a follow-up.

**Comment by thockin** [comment]
@pmorie which point do you want to reconsider?

File: pkg/kubelet/kubelet_pods.go [hunk 16]
```diff
@@ -414,6 +414,47 @@ func (kl *Kubelet) makeEnvironmentVariables(pod *v1.Pod, container *v1.Container
...
+					return result, fmt.Errorf("Invalid environment variable name, %v, from configmap %v/%v: %s", k, pod.Namespace, name, errMsgs[0])
+				}
+
+				if len(envFrom.Prefix) > 0 {
+					k = envFrom.Prefix + k
```
Comment (liggitt) [inline pkg/kubelet/kubelet_pods.go]: shouldn't the IsCIdentifier check go after we've prepended the prefix and determined the final `k` we will use?

Comment (fraenkel) [inline pkg/kubelet/kubelet_pods.go]: In theory it shouldn't matter but it can be moved.
The Prefix is already guaranteed to be a valid C identifier. I cannot imagine a case where you merge 2 c identifiers create an invalid c identifier. But it would be more clear.

Comment (liggitt) [inline pkg/kubelet/kubelet_pods.go]: > I cannot imagine a case where you merge 2 c identifiers create an invalid c identifier. But it would be more clear.

Other way around. A prefix could make a key that is not a valid C identifier (like `123`) valid (e.g. `FOO_123`).

Comment (fraenkel) [inline pkg/kubelet/kubelet_pods.go]: Will fix this in a separate PR. Thi is the case where a configmap can make a prefix required.

**Comment by zmerlynn** [comment]
@k8s-bot aws e2e test this

**Comment by zmerlynn** [comment]
@k8s-bot aws e2e test this
